### PR TITLE
[spirv] Migrate to perform vectorization on tensor ops 

### PR DESCRIPTION
**Pull request by antiagainst**
Previously the first thing we do after entering the SPIR-V CodeGen
backend is to bufferize. Then we deduce CodeGen configuration
and invoke the whole pipeline.

This commit moves the bufferization step downward. We first try
to deduce CodeGen configuration on tensor ops, and then invoke
the CodeGen pipeline. Notably for the vectorization pipeline,
vectorization now is done on Linalg ops with tensor semantics,
with the full flow of: tiling to invocations, vectorization,
vector optimizations, bufferizing, distributing to invocations,
and then perform loop canonicalization.

### CODE REVIEW

**Comment by antiagainst** [comment]
Depends on https://reviews.llvm.org/D113927 and https://github.com/google/iree/pull/7650.

**Comment by antiagainst** [comment]
Depends on https://reviews.llvm.org/D114376.

File: iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp [hunk 1]
```diff
@@ -94,6 +95,53 @@ void SPIRVLowerExecutableTargetPass::runOnOperation() {
...
   }
 
+  //===--------------------------------------------------------------------===//
+  // TODO: This is HACK to unblock vectorization on tensor-semantic ops. For
+  // padding, it would mean that we need to support tiling and distributing
+  // linalg.pad_tensor ops. It's quite involved and still ongoing. So for now
+  // fallback to bufferization first (which will turn it into a `linalg.copy`
+  // op) before even deducing the CodeGen configuration.
+  if (!passPipeline) {
+    OpPassManager bufferizePipeline(
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp:105]: I dont think we need to do this. On the CPU side the way this is handled is that you decide the workload per workgroup and pipeline to use purely based on the loop ranges of the tiled loop.

In any case this seems too convoluted to follow. Would it be possible to use a separate pipeline that bufferizes first.

Acknowledge that this is a hack to get things working. Do you have an ETA of when you can do away with this hack?

File: iree/compiler/Codegen/SPIRV/SPIRVTile.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  auto getTileSizeFn = [&](OpBuilder &builder, Operation *op) {
+    return getTileSizes(builder, op, 2);
+  };
+  auto tilingOptions = linalg::LinalgTilingOptions()
+                           .setLoopType(linalg::LinalgTilingLoopType::Loops)
+                           .setTileSizeComputationFunction(getTileSizeFn);
+  auto marker = Identifier::get(getTileReductionMarker(), context);
+  auto filter = linalg::LinalgTransformationFilter({marker});
+
+  patterns.insert<linalg::LinalgTilingPattern<linalg::BatchMatmulOp>,
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:46]: Note for future. We should move to just use the op agnostic patterns here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Awesome! Just one comment as you might have expected.

File: iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp]: Leftover from earlier prototyping. Moved to use a separate pipeline to handle padding/copy. This should be a better structure now.

File: iree/compiler/Codegen/SPIRV/SPIRVTile.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:46]: Yup. This should be relatively straightforward and we can do it later.

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 3]
```diff
@@ -649,19 +648,16 @@ LogicalResult initSPIRVLaunchConfig(ModuleOp module) {
...
-        const int64_t subgroupSize =
-            limits.subgroup_size().getValue().getSExtValue();
-        std::array<int64_t, 3> workgroupSize = {subgroupSize, 1, 1};
-        SmallVector<int64_t> workloadPerWorkgroup(tiledLoops.size(), 1);
-        workloadPerWorkgroup.front() = subgroupSize * 4;
-        setTranslationInfo(
-            funcOp,
-            IREE::Codegen::DispatchLoweringPassPipeline::SPIRVDistribute,
-            workloadPerWorkgroup, workgroupSize);
+        // TODO: This is HACK to unblock vectorization on tensor-semantic ops.
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/KernelConfig.cpp]: This part is rewritten so not applicable anymore.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Codegen/SPIRV/SPIRVTile.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+static void populateTilingReductionPatterns(RewritePatternSet &patterns) {
+  MLIRContext *context = patterns.getContext();
+  auto getTileSizeFn = [&](OpBuilder &builder, Operation *op) {
+    return getTileSizes(builder, op, 2);
+  };
+  auto tilingOptions = linalg::LinalgTilingOptions()
+                           .setLoopType(linalg::LinalgTilingLoopType::Loops)
+                           .setTileSizeComputationFunction(getTileSizeFn);
+  auto marker = Identifier::get(getTileReductionMarker(), context);
+  auto filter = linalg::LinalgTransformationFilter({marker});
```
Comment (ScottTodd) [inline iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:44]: This broke our gcc build (tested on post-submit). Logs here: https://buildkite.com/iree/iree-build-configurations/builds/1496#91c054cd-df7a-4758-84bd-91a621668581

error message:
```
[2021-12-02T19:07:52Z] ../iree/compiler/Codegen/SPIRV/SPIRVTile.cpp: In function 'void mlir::iree_compiler::populateTilingReductionPatterns(mlir::RewritePatternSet&)':
[2021-12-02T19:07:52Z] ../iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:44:60: error: call of overloaded 'LinalgTransformationFilter(<brace-enclosed initializer list>)' is ambiguous
[2021-12-02T19:07:52Z]    44 |   auto filter = linalg::LinalgTransformationFilter({marker});
[2021-12-02T19:07:52Z]       |                                                            ^
[2021-12-02T19:07:52Z] In file included from ../iree/compiler/Codegen/Passes.h:15,
[2021-12-02T19:07:52Z]                  from ../iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:15:
[2021-12-02T19:07:52Z] ../third_party/llvm-project/llvm/../mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:436:3: note: candidate: 'mlir::linalg::LinalgTransformationFilter::LinalgTransformationFilter(const mlir::linalg::LinalgTransformationFilter&)'
[2021-12-02T19:07:52Z]   436 |   LinalgTransformationFilter(const LinalgTransformationFilter &) = default;
[2021-12-02T19:07:52Z]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
[2021-12-02T19:07:52Z] ../third_party/llvm-project/llvm/../mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:435:3: note: candidate: 'mlir::linalg::LinalgTransformationFilter::LinalgTransformationFilter(mlir::linalg::LinalgTransformationFilter&&)'
[2021-12-02T19:07:52Z]   435 |   LinalgTransformationFilter(LinalgTransformationFilter &&) = default;
[2021-12-02T19:07:52Z]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~
[2021-12-02T19:07:52Z] ../third_party/llvm-project/llvm/../mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:427:12: note: candidate: 'mlir::linalg::LinalgTransformationFilter::LinalgTransformationFilter(llvm::ArrayRef<mlir::StringAttr>, llvm::Optional<mlir::StringAttr>)'
[2021-12-02T19:07:52Z]   427 |   explicit LinalgTransformationFilter(
[2021-12-02T19:07:52Z]       |            ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (ScottTodd) [reply iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:44]: Sent a fix: https://github.com/google/iree/pull/7802

Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:44]: Ah, sorry about this. Thanks for the fix!!

**Pull request merged**